File: A2Z_striver/1 Basics/1.6 String/isomorphicStrings/isomorphicStrings.py
class Solution:
    # ~ Approach 1: Two Hash Maps (Most Readable)
    # Analysis: Time O(N) | Space O(N)
    def isomorphicStrings(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
            
        map_s_t = {} # Maps s -> t
        map_t_s = {} # Maps t -> s
        
        for c1, c2 in zip(s, t):
            # Case 1: c1 is seen before. Check consistency.
            if c1 in map_s_t:
                if map_s_t[c1] != c2:
                    return False
            
            # Case 2: c2 is seen before. Check consistency.
            if c2 in map_t_s:
                if map_t_s[c2] != c1:
                    return False
            
            # Case 3: New mapping. Add to both maps.
            map_s_t[c1] = c2
            map_t_s[c2] = c1
            
        return True

    # ~ Approach 2: One Map + One Set (Slightly faster in Python)
    # This fixes your original logic by using a Set for O(1) lookups
    def isomorphicStringsOptimized(self, s: str, t: str) -> bool:
        if len(s) != len(t): return False
        
        mapping = {}
        used_chars = set() # Replaces checking c.values()
        
        for c1, c2 in zip(s, t):
            if c1 in mapping:
                # If mapped, must match current char in t
                if mapping[c1] != c2:
                    return False
            else:
                # If not mapped, ensure c2 isn't already used by someone else
                if c2 in used_chars:
                    return False
                
                mapping[c1] = c2
                used_chars.add(c2)
                
        return True

# --- Driver Code ---
if __name__ == "__main__":
    sol = Solution()
    print(sol.isomorphicStrings("egg", "add"))  # True
    print(sol.isomorphicStrings("foo", "bar"))  # False
    print(sol.isomorphicStrings("paper", "title")) # True

File: A2Z_striver/1 Basics/1.6 String/isomorphicStrings/isomorphicStrings.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    // ~ Approach 1: Generic Map (Good for any characters)
    // Analysis: Time O(N) | Space O(N)
    bool isIsomorphicMap(string s, string t) {
        if (s.length() != t.length()) return false;

        unordered_map<char, char> map_s_t;
        unordered_map<char, char> map_t_s;

        for (int i = 0; i < s.length(); i++) {
            char c1 = s[i];
            char c2 = t[i];

            // Check s -> t mapping
            if (map_s_t.count(c1) && map_s_t[c1] != c2) return false;
            
            // Check t -> s mapping
            if (map_t_s.count(c2) && map_t_s[c2] != c1) return false;

            // Create mapping if neither existed
            map_s_t[c1] = c2;
            map_t_s[c2] = c1;
        }
        return true;
    }

    // ~ Approach 2: ASCII Arrays (Fastest for Interview)
    // Analysis: Time O(N) | Space O(1) (Fixed size 256)
    bool isIsomorphicArray(string s, string t) {
        if (s.length() != t.length()) return false;
        
        // Use arrays to store the last seen position + 1 of each char
        int m1[256] = {0};
        int m2[256] = {0};
        
        for (int i = 0; i < s.length(); i++) {
            // Logic: If 'e' in 's' was last seen at index 5, 
            // then 'a' in 't' must also have been last seen at index 5.
            if (m1[s[i]] != m2[t[i]]) return false;
            
            // Store index + 1 (to distinguish from default 0)
            m1[s[i]] = i + 1;
            m2[t[i]] = i + 1;
        }
        return true;
    }
};

int main() {
    Solution sol;
    cout << boolalpha;
    cout << "Test 1: " << sol.isIsomorphicArray("egg", "add") << endl;    // true
    cout << "Test 2: " << sol.isIsomorphicArray("foo", "bar") << endl;    // false
    return 0;
}